rfcs/003-strings.md

```diff
@@ -0,0 +1,325 @@
+}
+
+// This allows `CWString` to be turned into a `Borrowed<'a, PCWSTR>`
+impl From<&CWString> for &PCWSTR {}
+impl From<CWString> for HSTRING {}
+impl From<HSTRING> for CWString {}
+
+impl Drop for CWString {
+    // The `CWString` owns its memory and will free when its dropped
+}
+// Display shows the string with non-valid utf16 replaced with �
+impl Display for CWSTRING {}
+// Same as Display but surrounded by ""
+impl Debug for CWSTRING {}
+
+// HSTRING should also include implementations of `PartialEq` for 
+// `String`, `&str`, `OsString`, and `&OsStr`
+
+// --------------------------------
+
+impl BSTR {
+    fn from_ptr(ptr: *const c_void) -> Self {}
+    fn as_ptr(&self) -> *const c_void;
+    // String data without the trailing 0
+    unsafe fn as_wide(&self) -> &[u16] {}
+    // String data with the trailing 0
+    unsafe fn as_wide_with_nul(&self) -> &[u16] {}
+    unsafe fn to_string(&self) -> Result<String, FromUtf16Error> {}
+    unsafe fn to_string_lossy(&self) -> String {}
+}
+
+
+// --------------------------------
+
+impl PCWSTR {
+    fn from_ptr(ptr: *const u16) -> Self {}
+    fn as_ptr(&self) -> *const u16;
+    fn is_null(&self) -> bool {}
+    // String data without the trailing 0
+    unsafe fn as_wide(&self) -> &[u16] {}
+    // String data with the trailing 0
+    unsafe fn as_wide_with_nul(&self) -> &[u16] {}
+    unsafe fn to_string(&self) -> Result<String, FromUtf16Error> {}
+    unsafe fn to_string_lossy(&self) -> String {}
+    unsafe fn display(&self) -> impl Display {}
+}
+// This allows `CString` and `CSTR` to be turned into a `Borrowed<'a, PCSTR>`
+impl From<&CString> for &PCSTR {}
+impl From<&CStr> for &PCSTR {}
+
+// --------------------------------
+
+impl PCSTR {
+    fn from_ptr(ptr: *const u8) -> Self {}
+    fn as_ptr(&self) -> *const u8;
+    fn is_null(&self) -> bool {}
+    // String data without the trailing 0
+    unsafe fn as_bytes(&self) -> &[u8] {}
+    // String data with the trailing 0
+    unsafe fn as_bytes_with_nul(&self) -> &[u8] {}
+    // Converts to `&str` checking for valid UTF-8
+    unsafe fn as_str(&self) -> Result<&str, FromUtf8Error> {}
+    // Converts to `&str` not checking for valid UTF-8
+    unsafe fn as_str_unchecked(&self) -> &str {}
+}
+
+// Note that `Display`, `Debug`, and the conversion traits are not 
+// included for non-`HSTRING` types because those traits are not 
+// marked as unsafe
+```
+
+#### String literals
+
+Many times the user simply wants to use a string literal to build a string of a certain type.
+
+The windows crate provides the following macros for convenience:
+
+```rust
+let x: CWString = w!("hello");
+let y: CString = c!("hello");
+let z: HSTRING = h!("hello");
+```
+
+These function build null terminated string data of the appropriate width (`u8` in the case of `PCSTR` and `u16` in the other cases).
+
+**question**: for all these types built from static strings, could we special case them so that they don't use reference counting at all?
```

Comment:
You could just special case for a string literal, then use that with `concat!($str, "\0")`, etc. For wide strings, you'd probably need something like [`const_heap`](https://github.com/rust-lang/rust/issues/79597) (nightly api very far from stable) or compiler support.

Reply:
This refers to whether we need a special version of `HSTRING` and the like that doesn't track reference counts since its backing memory is statically allocated. Actually statically allocating enough memory for a wide string is already a solved problem (e.g., [see here](https://github.com/rylev/const-utf16)).

Reply:
Ok. I didn't realise that the const on stable had got that far.

Reply:
So essentially, you'd get a wide string reference (`PCWSTR`), then pass that to [`WindowsCreateStringReference`](https://docs.microsoft.com/en-us/windows/win32/api/winstring/nf-winstring-windowscreatestringreference) ([in `windows::Win32::System::WinRT`](https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/System/WinRT/fn.WindowsCreateStringReference.html)), then use the resultant `HSTRING` structure.

Reply:
Using `WindowsCreateStringReference` is generally a pessimization unless the calling language natively supports null terminated wide string literals, like C# and C++. For languages like Rust, that's not worth it as the resultant `HSTRING` typically ends up being copied almost immediately.

Reply:
Then we shouldn't need the `Into<T>` below, unless I'm forgetting something.